server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -269,6 +270,16 @@ public NodeEnvironment(Settings settings, Environment environment) throws IOExce
                 assertCanWrite();
             }
 
+            // versions 7.x and earlier put their data under ${path.data}/nodes/; leave a file at that location to prevent downgrades
+            final Path legacyNodesPath = environment.dataFile().resolve("nodes");
+            if (Files.isRegularFile(legacyNodesPath) == false) {
+                final String content = "written by Elasticsearch v" + Version.CURRENT +
```

Comment:
I wonder if we should follow-up with a 7.16 change to log the first 256 bytes of the file if `nodes` is a file and not a directory?

Reply:
heh yeah I suppose we could at least help 7.16 users here :)